pkg/forwarder/forwarder.go

```diff
@@ -346,3 +383,87 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
 	return f.sendHTTPTransactions(transactions)
 }
+
+// SubmitProcessChecks sends process checks
+func (f *DefaultForwarder) SubmitProcessChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakeProcesses.Add(1)
+
+	return f.submitProcessLikePayload(processesEndpoint, payload, extra, true)
+}
+
+// SubmitRTProcessChecks sends real time process checks
+func (f *DefaultForwarder) SubmitRTProcessChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakeRTProcesses.Add(1)
+
+	return f.submitProcessLikePayload(rtProcessesEndpoint, payload, extra, false)
+}
+
+// SubmitContainerChecks sends container checks
+func (f *DefaultForwarder) SubmitContainerChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakeContainer.Add(1)
+
+	return f.submitProcessLikePayload(containerEndpoint, payload, extra, true)
+}
+
+// SubmitRTContainerChecks sends real time container checks
+func (f *DefaultForwarder) SubmitRTContainerChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakeRTContainer.Add(1)
+
+	return f.submitProcessLikePayload(rtContainerEndpoint, payload, extra, false)
+}
+
+// SubmitConnectionChecks sends connection checks
+func (f *DefaultForwarder) SubmitConnectionChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakeConnections.Add(1)
+
+	return f.submitProcessLikePayload(connectionsEndpoint, payload, extra, true)
+}
+
+// SubmitPodChecks sends pod checks
+func (f *DefaultForwarder) SubmitPodChecks(payload Payloads, extra http.Header) (chan Response, error) {
+	transactionsIntakePod.Add(1)
+
+	return f.submitProcessLikePayload(podEndpoint, payload, extra, true)
+}
+
+func (f *DefaultForwarder) submitProcessLikePayload(ep endpoint, payload Payloads, extra http.Header, retryable bool) (chan Response, error) {
+	transactions := f.createHTTPTransactions(ep, payload, false, extra)
+
+	results := make(chan Response, len(transactions))
+
+	var wg sync.WaitGroup
+	for _, txn := range transactions {
+		wg.Add(1)
+
+		txn.retryable = retryable
+		txn.attemptHandler = func(transaction *HTTPTransaction) {
+			if v := transaction.Headers.Get("X-DD-Agent-Timestamp"); v == "" {
+				transaction.Headers.Set("X-DD-Agent-Timestamp", strconv.Itoa(int(transaction.GetCreatedAt().Unix())))
+			}
+
+			if v := transaction.Headers.Get("X-DD-Agent-Attempts"); v == "" {
+				transaction.Headers.Set("X-DD-Agent-Attempts", "1")
+			} else {
+				attempts, _ := strconv.ParseInt(v, 10, 0)
+				transaction.Headers.Set("X-DD-Agent-Attempts", strconv.Itoa(int(attempts+1)))
+			}
+		}
+
+		txn.completionHandler = func(transaction *HTTPTransaction, statusCode int, body []byte, err error) {
+			results <- Response{
+				Domain:     transaction.Domain,
+				Body:       body,
+				StatusCode: statusCode,
+				Err:        err,
+			}
+			wg.Done()
+		}
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
```

Comment:
This will create a goroutine by transaction currently being queued. We might drop a transaction without calling the completion handler for a number of reasons ([retry queue is fulled](https://github.com/DataDog/datadog-agent/blob/master/pkg/forwarder/domain_forwarder.go#L108), the domain was black listed, ...).

I don't think we should rely on it being closed or add a dropped handler and call it from everywhere. I'm less in favor of the second as it's so easy to forget to call it in future development leading to a major leak in the agent.

Reply:
Good point, I will rewrite this so that there is not a reliance on the channel being closed.